### DISCUSSION TITLE
Missing fields in DataFetchingEnvironment

### DISCUSSION THREAD

**Discussion by FidasL** [ExpediaGroup/graphql-kotlin#1444] [Q&A]
Hello
I am not sure if this is question for you or better for graphql-java. But let's have this example of GraphQL Resolver implementation:
```
@GraphQLDescription("An object with globally unique ID")
interface Node {
    val nodeID: ID
}

data class PersonNode(override val nodeID: ID, val age: Int, val address: String): Node
data class AnimalNode(override val nodeID: ID, val name: String): Node


class MyQuery {
    fun node(env: DataFetchingEnvironment, id: ID): Node {
        return if (decodeId(id) == something) {
            ...fetch object by id query param

            AnimalNode(id, "my dog's name")
        } else {
            ...fetch object by id query param

            PersonNode(id, 10, "address")
        }
    }
}
```

If I call query like this:
```
query {
  node(id: "base64") {
    ... on PersonNode {
      address
      age
    }
  }
}
```
when calling `env.selectionSet.fields` in `node()` function, I am not able to get specified fields from graphql request. It is important in my use-case, because I want to query from database only  those columns which client specified..  My question is if there is any way to   distinguish which fields client wants to query?

**Comment by samuelAndalon** [comment]
Hello, what is the result of `env.selectionSet.fields` when you call it in your function ?
based on the documentation of `graphql-java` the `DataFetchingFieldSelectionSet` does actually takes the fields from fragments and "flatten" them into a list 

check https://github.com/graphql-java/graphql-java/blob/master/src/test/groovy/graphql/schema/DataFetchingFieldSelectionSetImplTest.groovy#L452

**Reply by FidasL** [reply]
It is an empty list (see image bellow). But I have got older versions -> `graphql-kotlin:3.7.0` and `graphql-java:15.0` . Not sure if there weren't some bugs.. 
![image](https://user-images.githubusercontent.com/32110313/167803586-751a32db-7c03-4a6a-93bc-be6ed989daad.png)

**Reply by samuelAndalon** [reply]
yes, you are using a very old version of both graphql-java and graphql-kotlin, try updating graphql-kotlin to latest

**Reply by FidasL** [reply]
It is not so easy to bump versions in our large graphql project. I just wasn't sure why selectionSet doesn't contain selected fields if I use interface in Graphql schema. If I query data directly for specific class, selectionSet contains fields specified in request.

**Reply by dariuszkuc** [reply]
By "interface" you mean fragments? Unsure about the `DataFetchingFieldSelectionSet` but regular selection set makes a distinction between `Fields` and `FragmentSpread`/`InlineFragment`. You could check with https://github.com/graphql-java/graphql-java whether they have any insights.

That being said those lib versions are very old and unsupported anymore so would highly recommend to look into update path.
